smolyay/samples.py

```diff
@@ -0,0 +1,158 @@
+    ``number_points`` is the number of points that the UnidimensionalPointSet
+    is expected to store.
+    
+    Parameters
+    ----------
+    natural_domain : list of two numbers
+        the natural domain of the function within which calls are made.
+    """
+
+    def __init__(self,natural_domain):
+        self._natural_domain = natural_domain
+        self._points = []
+        self._number_points = 0
+
+    @property
+    def natural_domain(self):
+        """list: Domain the sample points come from."""
+        return self._natural_domain
+    
+    @property
+    def points(self):
+        """list: Points stored by the set."""
+        return self._points
+
+    @property
+    def number_points(self):
+        """int: The number of points stored by the set."""
+        return self._number_points
+    
+    @number_points.setter
+    def number_points(self, value):
+        if not isinstance(value, int):
+            raise TypeError('Number must be integer greater than 0.')
+        if value <= 0:
+            raise ValueError('Number of points must be greater than 0')
+        if len(self._points) < value:
+            self._points = self._generate_points(value)
+        self._points = self._points[:value]
+        self._number_points = len(self._points)
+
+    @abc.abstractmethod
+    def _generate_points(self, num_points):
+        r"""Generating the points
+
+        An abstract method for generating the points stored by the point set.
+        Called when the number of points requested is larger than the number
+        of points available.
+
+        Parameters
+        ----------
+        x : int
+            The number of points to generate
+
+        Returns
+        -------
+        list
+            points stored by the set
+        """
+        pass
+
+
+class ClenshawCurtisPointSet(UnidimensionalPointSet):
+    r"""Set of unidimensional points for Clenshaw Curtis sampling
+
+    The :attr:`points` for this interpolation scheme are the extrema of the 
+    Chebyshev polynomials of the first kind on the domain :math:`[-1, 1]`:
+
+    .. math::
+
+        x_i^* = -\cos(\pi i/n), i = 0,...,n
+
+    For the special case :math:`n = 0`, there is only one point :math:`x_0^* = 0`.
+    """
+
+    def __init__(self):
+        super().__init__([-1, 1])
+
+    def _generate_points(self, num_points):
+        r"""Generating the points
+
+        Parameters
+        ----------
+        x : int
+            The number of points to generate
+
+        Returns
+        -------
+        list
+            Value of Chebyshev polynomial of the first kind.
+        """
+        print('here')
+        points = [0]
+        degree = 0
+        counter = 0
+        while len(points) < num_points:
+            counter = counter + 1
+            degree = 2**counter
+            new_points = -numpy.cos(numpy.pi * numpy.linspace(0, degree, degree + 1) / degree)
+            for p in new_points:
```

Comment:
In theory, it should be possible to use numpy matrix math and the equation used by `numpy.isclose` to get a boolean list that shows exactly which points in `new_points` are new. However, I don't know if doing that would be faster.

Reply:
Given the numerical difficulties we've had with this, I think we should use integer math to figure out which new points are being added at each level so this is exact. We can then convert the integers to points once they are unique.

However, I have a question: can't the Clenshaw-Curtis points be defined for arbitrary N, and not require this nested formalism? I think we might want to come up with a way to distinguish this, and that perhaps includes setting levels. Like, maybe the default constructor here just takes an `N` and makes the normal CC points.

Then, we can make a nested version that derives from this according to the various scheme, and that takes an exactness / level in its constructor. Kinda similar to how we were using the classmethod in the old version.

Reply:
I used the greatest common factor to prune out the duplicate points before they were calculated.

Clenshaw-Curtis points can be defined by any of the basis functions that have the nested extrema. The order is just the order that the default exponential growth uses.

I have been thinking about potentially having different orders. The Clenshaw-Curtis Linear growth that adds 2 points on every level besides the first can't use the exponential growth order because then the points added every level wouldn't be symmetrical. If you remove all the positive points like this:

[0, -1.0, -0.7071067811865476, -0.9238795325112867, -0.38268343236508984]

you can see an "essential" order that the points are added in. Then you can decide where you want the level boundaries to be and add the positive values to make all the levels symmetrical.

So while I've never seen the Clenshaw-Curtis Linear growth, I imagine it divides things up like this:

[0] [-1.0, 1.0] [-0.7071067811865476, 0.7071067811865476] [-0.9238795325112867, 0.9238795325112867][-0.38268343236508984, 0.38268343236508984]

Reply:
I think I see what you’re saying, but I was thinking that you can define these points even for a highest degree polynomial that is non-nested?

https://en.m.wikipedia.org/wiki/Clenshaw%E2%80%93Curtis_quadrature

Having these points would be important, for example, to get access to more tensor grids other than those that match the nested criteria.

Reply:
Would you want that to be the generate_points method and then the Clenshaw-Curtis has children that do use levels override it, or do you want generating the points in a non-nested order to be a separate method?

Reply:
I've added a child to the ClenshawCurtisPointSet called the NestedClenshawCurtisPointSet, which uses the generate_points that its parent used to have. The ClenshawCurtisPointSet now generates N extrema for a polynomial of degree N - 1. 

I've also gotten rid of the number_points property so now the user is expected to make points using the generate_points method.